Fix forums sorting issues

miankhalid opened the pull request:
Fixes [MA-2163](https://openedx.atlassian.net/browse/MA-2163)

I've cut-down the PR into 3 commits, each with very verbose :wink: commit message of what its for, making it easy to review.

@1zaman @bguertin @BenjiLee plz review.

1zaman commented on VideoLocker/src/main/java/org/edx/mobile/view/adapters/BaseListAdapter.java:
```diff
@@ -125,6 +125,15 @@ public void replace(T item, int position) {
+        if (itemPosition < objects.size()) {
+            T item = objects.get(itemPosition);
+            objects.remove(itemPosition);
+            objects.add(0, item);
```
This can be replaced with:

````java
Collections.swap(objects, 0, itemPosition);
````

And if you're validating the index, then you should validate against the minimum (zero) as well.

1zaman replied:
In fact instead of the specific `moveToTop()` method, you might define a generic `swapItems()` method that takes two indices explicitly. It's up to you.

miankhalid replied:
I actually swapped the items initially, but thats wrong logically, cuz we don't want 2 items to exchange places.

1zaman replied:
:confounded: Looks like I skimmed too fast.

1zaman commented on VideoLocker/src/main/java/org/edx/mobile/view/CourseDiscussionPostsThreadFragment.java:
```diff
@@ -171,6 +172,24 @@ public void onDestroy() {
+        if (postsSort == DiscussionPostsSort.LAST_ACTIVITY_AT && mSelectedItem != -1) {
+            // Move the last viewed thread to the top
+            discussionPostsAdapter.moveToTop(mSelectedItem);
+            mSelectedItem = -1;
```
This can be done directly from the `onItemClick()` callback, thus removing the need to store the click index. Or, if this change is somehow visible through the new Activity transition, then this action can be posted on the event queue (i.e. via a `Handler`) from there.

miankhalid replied:
The post is immediately moved to the top before the activity transition, even if we post it on the handler. So I guess, current approach is correct.

1zaman commented on VideoLocker/src/main/java/org/edx/mobile/view/CourseDiscussionPostsThreadFragment.java:
```diff
@@ -171,6 +172,24 @@ public void onDestroy() {
+        }
+
+        nextPage = 1;
+        controller.refresh();
```
The implementation of the `onRestart()` callback is not appropriate here, as we only want to reload the list on the event of the user returning from the item detail Activity, not in the case of any and all restarts. Therefore, I guess we can do this from the `onActivityResult()` callback, and start the details Activity via the `startActivityForResult()` method with a non-negative request code.

miankhalid replied:
For now this seems to be a better option, cuz, we can't reload the thread list forcefully. So, silently updating the list ensures a user can always see the newly added posts whenever he/she comes back to this screen.

1zaman replied:
Makes sense, I guess. We should add pull to refresh functionality at some point though, at which point my suggested change can be implemented.

1zaman commented on VideoLocker/src/main/java/org/edx/mobile/view/adapters/InfiniteScrollUtils.java:
```diff
@@ -124,14 +125,32 @@ public void loadMore() {
+         * @param lazyClearanceEnabled <code>true</code> if we want to clear after the page has
+         *                             loaded, otherwise <code>false</code>.
+         */
+        private void onLoadMore(boolean showProgress, final boolean lazyClearanceEnabled) {
```
Since the two parameters always have the opposite value, it might make sense to condense them to just one here.

miankhalid replied:
seems logical.

1zaman commented on VideoLocker/src/main/java/org/edx/mobile/view/adapters/InfiniteScrollUtils.java:
```diff
@@ -154,11 +173,21 @@ private boolean isAbandoned() {
+        }
+
+        @Override
+        public void refresh() {
```
I think naming this `resetSilently()` would be more accurate, since it _is_ a destructive operation (discarding the pagination). But the final naming is up to you.

miankhalid replied:
ok.

1zaman commented on VideoLocker/src/main/java/org/edx/mobile/view/CourseDiscussionPostsThreadFragment.java:
```diff
@@ -197,15 +216,15 @@ public void onEventMainThread(DiscussionCommentPostedEvent event) {
             }
-            discussionPostsAdapter.insert(event.getDiscussionThread(), i);
+
+            discussionPostsAdapter.insert(newThread, 0);
```
Hmm, actually the previous implementation seems fine here. We shouldn't be adding new threads on top of pinned threads.

miankhalid replied:
yeah you are right.

1zaman commented on VideoLocker/res/values/strings.xml:
```diff
@@ -296,6 +296,10 @@
+    <!-- Label for letting the user know his/her response was successfully posted -->
+    <string name="discussion_response_posted">Your response is now posted</string>-->
+    <!-- Label for letting the user know his/her comment was successfully posted -->
+    <string name="discussion_comment_posted">Your comment is now posted</string>
```
These should have period at the end, I guess. Get final wording from docs team?

miankhalid replied:
We are waiting on the docs team currently for the correct strings [here](https://openedx.atlassian.net/browse/MA-2162).

1zaman commented on VideoLocker/src/main/java/org/edx/mobile/view/CourseDiscussionCommentsFragment.java:
```diff
@@ -159,8 +162,12 @@ public void onDestroy() {
-            discussionCommentsAdapter.insertCommentAtEnd(event.getComment());
+        if (!hasMorePages) {
+            if (null != event.getParent() && event.getParent().getIdentifier().equals(discussionResponse.getIdentifier())) {
+                ((BaseFragmentActivity)getActivity()).showInfoMessage(getString(R.string.discussion_comment_posted));
```
The message should be shown even if there are more pages to load yet. In fact, that is the scenario where it really needs to be shown :stuck_out_tongue: If the item is displayed, then we could even dispense with it.

miankhalid replied:
Good catch. Tried to be over-smart I guess :wink: .

1zaman commented on VideoLocker/src/main/java/org/edx/mobile/view/CourseDiscussionResponsesFragment.java:
```diff
@@ -136,9 +138,15 @@ public void onEventMainThread(DiscussionCommentPostedEvent event) {
+                }
             } else {
                 // We got a comment to a response
+                ((BaseFragmentActivity)getActivity()).showInfoMessage(getString(R.string.discussion_comment_posted));
```
I think we don't need to show the new comments posting notification in this screen, as it's not directly interacting with it, and we aren't currently showing it globally.

miankhalid replied:
There is a case when we need to show this message i.e. when a response has no comments on it.

1zaman replied:
In that case I guess this behaviour is fine.

1zaman commented:
Overall :+1:

miankhalid commented:
All comments addressed. Care for a 2nd pass @1zaman ?

p.s. I've addressed issues of each commit in a separate new commit prefixed with commit number of the commit its addressing the feedback i.e. [`C1:Refactoring`](f855573) addresses feedback for [Commit # 1](cdd4dc9).

1zaman commented on VideoLocker/src/main/java/org/edx/mobile/view/adapters/InfiniteScrollUtils.java:
```diff
@@ -124,14 +125,31 @@ public void loadMore() {
+
+        /**
+         * Variant of the {@link #onLoadMore} function allowing us to control the visibility of
+         * progress and lazily clear adapter after a page has loaded.
```
Nitpick: this is the main method, not a variant :stuck_out_tongue: I suggest either remove the description, or have it be the same as the no-args one.

1zaman commented on VideoLocker/src/main/java/org/edx/mobile/view/adapters/InfiniteScrollUtils.java:
```diff
@@ -124,14 +125,31 @@ public void loadMore() {
+         *
+         * @param lazyClearanceEnabled <code>true</code> if we want to clear the adapter after the
+         *                             page has loaded without showing progress, otherwise
+         *                             <code>false</code>.
```
Alternative name and description, that is possibly more concise and clear as to the intent:

````java
/**
  * @param isRefreshingSilently <code>true</code> If we're doing a silent refresh,
  *                             <code>false</code> if pagination is being done as usual.
  */
````

1zaman commented:
:+1:

miankhalid commented:
Just need a 2nd review and strings from the doc team.

miankhalid commented on VideoLocker/res/values/strings.xml:
```diff
@@ -296,10 +296,10 @@
+    <!-- Label for letting the user know his/her response was successfully added -->
+    <string name="discussion_response_posted">Your response has been added.</string>-->
+    <!-- Label for letting the user know his/her comment was successfully added -->
+    <string name="discussion_comment_posted">Your comment has been added.</string>
```
Strings updated. @catong plz review.

miankhalid commented:
Strings have been updated as advised by the docs team.
We just need a 2nd review from one of u @BenjiLee | @bguertin to merge this PR.

BenjiLee commented:
When adding a the first comment to a response, instead of staying on the response screen, it takes me back to the post. When adding a second response, it will take me back to the post screen. I think the latter is the expected behaviour.

BenjiLee commented:
Other than that comment :+1:

catong commented:
@miankhalid Strings look good, thanks! :+1:

miankhalid commented:
> So when I'm adding a comment, and it's the first comment to the response, after I add the comment I am taken back to the post screen where it lists the responses, instead of the response screen where it lists the comments

Yeah, that is sort of redirection problem that we missed. We should create separate tickets for both platforms cuz iOS does the same.

As far as this PR is concerned, the fore mentioned requirement is out of scope, which is the reason why this PR has been merged.

BenjiLee commented:
https://openedx.atlassian.net/browse/MA-2191 Forgot to post the ticket!

miankhalid commented:
ohh .. Mubashar created tickets as well 1 for iOS and another for android. Let me take care of the clones :wink:
